pylinalg/obj/transform.py

```diff
@@ -0,0 +1,137 @@
+import numpy as np
+
+from ..func import quaternion as qt
+
+
+class Transform:
+    """
+    Abstract base class to define a common interface and allow future extension
+    to non-linear transformations.
+    """
+
+
+class AffineTransform(Transform):
+    """
+    A affine transformation.
+
+    This transform represents an affine mapping between two coordinate frames:
+    parent and child. That is, a transformation that is linear in homogeneous
+    coordinates and that consists of a translation, rotation, and scale (applied
+    in that order).
+
+
+    Parameters
+    ----------
+    position : ndarray, [3]
+        The position of the child's frame in the parent's space. (The
+        translation to apply to map vectors from child to parent.)
+    orientation : ndarray, [4]
+        The orientation of the parent's frame relative to the child's frame. (A
+        quaternion (format: x, y, z, w) describing the rotation from child to
+        parent.)
+    scale : ndarray, [3]
+        The scale of the parent expressed in child's units. (A axis-wise scaling
+        to map units from child to parent.)
+
+    """
+
+    # some constants for syntax sugar
+    FORWARD = np.array((0, 0, 1))
+    BACKWARD = np.array((0, 0, -1))
+    UP = np.array((0, 1, 0))
+    DOWN = np.array((0, -1, 0))
+    RIGHT = np.array((1, 0, 0))
+    LEFT = np.array((-1, 0, 0))
+
+    def __init__(self, /, *, position=None, orientation=None, scale=None) -> None:
+        self._position = position
+        self._orientation = orientation
+        self._scale = scale
+
+        if position is None:
+            self._position = np.zeros(3)
+
+        if orientation is None:
+            self._orientation = np.array((0, 0, 0, 1))
```

Comment:
I think you may want to ensure this is an f8 ndarray like the other two, to avoid extra work casting.